agent/command.go

```diff
@@ -165,15 +165,19 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
 	}
 	defer func() {
 		if !tc.unsetFunctionVarsDisabled || tc.project.UnsetFunctionVars {
-			// This defer ensures that the function vars do not persist in the expansions after the function is over.
-			for key, functionValue := range commandInfo.Vars {
-				currentValue := tc.taskConfig.Expansions.Get(key)
-				if currentValue != functionValue {
-					// If a command in the func updates the expansion value, don't reset it.
-					delete(prevExp, key)
+			// This defer ensures that the function vars do not persist in the expansions after the function is over
+			// unless they were updated using expansions.update
+			if cmd.Name() == "expansions.update" && tc.taskConfig.DynamicExpansions != nil {
```

Comment:
true, I'll add a length check. But I'd also rather keep both checks because there are no guarantees that DynamicExpansions won't be used in some other way in the future.